commit 3d360b97ed7dead8e9a5f602cfaf61ab7b41e531
Author: Venky Shankar <vshankar@redhat.com>
Date:   Mon Oct 14 06:51:03 2019 -0400

    mgr: invoke plugin shutdown on SIG{TERM,INT} signals.
    
    Mgr standby used to install signals handler but did pretty
    much nothing when signals were received. Move the signal
    handling part to Mgr core so as to invoke plugin shutdowns
    when ceph-mgr received SIGTERM or SIGINT.
    
    Fixes: http://tracker.ceph.com/issues/42299
    Signed-off-by: Venky Shankar <vshankar@redhat.com>

diff --git a/src/mgr/Mgr.cc b/src/mgr/Mgr.cc
index 27f74f2b06e..a10746a2cfb 100644
--- a/src/mgr/Mgr.cc
+++ b/src/mgr/Mgr.cc
@@ -219,12 +219,35 @@ std::map<std::string, std::string> Mgr::load_store()
   return loaded;
 }
 
+void Mgr::handle_signal(int signum)
+{
+  ceph_assert(signum == SIGINT || signum == SIGTERM);
+  derr << "*** Got signal " << sig_str(signum) << " ***" << dendl;
+  shutdown();
+  _exit(0);  // exit with 0 result code, as if we had done an orderly shutdown
+}
+
+// A reference for use by the signal handler
+static Mgr *signal_mgr = nullptr;
+
+static void handle_mgr_signal(int signum)
+{
+  if (signal_mgr) {
+    signal_mgr->handle_signal(signum);
+  }
+}
+
 void Mgr::init()
 {
   std::unique_lock l(lock);
   ceph_assert(initializing);
   ceph_assert(!initialized);
 
+  // Enable signal handlers
+  signal_mgr = this;
+  register_async_signal_handler_oneshot(SIGINT, handle_mgr_signal);
+  register_async_signal_handler_oneshot(SIGTERM, handle_mgr_signal);
+
   // Start communicating with daemons to learn statistics etc
   int r = server.init(monc->get_global_id(), client_messenger->get_myaddrs());
   if (r < 0) {
@@ -395,6 +418,7 @@ void Mgr::load_all_metadata()
 
 void Mgr::shutdown()
 {
+  dout(10) << "mgr shutdown init" << dendl;
   finisher.queue(new LambdaContext([&](int) {
     {
       std::lock_guard l(lock);
diff --git a/src/mgr/Mgr.h b/src/mgr/Mgr.h
index 85b1137fe0f..d38a75cac93 100644
--- a/src/mgr/Mgr.h
+++ b/src/mgr/Mgr.h
@@ -94,6 +94,8 @@ public:
   void background_init(Context *completion);
   void shutdown();
 
+  void handle_signal(int signum);
+
   std::map<std::string, std::string> get_services() const;
 };
 
